pkg/cmd/certregenerationcontroller/certregenerationcontroller.go

```diff
@@ -0,0 +1,295 @@
+	cmd.PersistentFlags().StringVarP(&o.TLSServerName, "tls-server-name", "", o.TLSServerName, "The SNI hostname to set for the server in kubeconfig")
+
+	cmd.PersistentFlags().DurationVar(&o.LeaderelectionLeaseDuration, "leaderelection-lease-duration", o.LeaderelectionLeaseDuration, "LeaseDuration is the duration that non-leader candidates will wait to force acquire leadership.")
+	cmd.PersistentFlags().DurationVar(&o.LeaderelectionRenewDeadline, "leaderelection-renew-deadline", o.LeaderelectionRenewDeadline, "RenewDeadline is the duration that the acting master will retry refreshing leadership before giving up.")
+	cmd.PersistentFlags().DurationVar(&o.LeaderelectionRetryPeriod, "leaderelection-retry-period", o.LeaderelectionRetryPeriod, "RetryPeriod is the duration the LeaderElector clients should wait between tries of actions.")
+
+	return cmd
+}
+
+func (o *Options) Validate() error {
+	return nil
+}
+
+func (o *Options) Complete() error {
+	var err error
+
+	if len(o.KubeconfigPath) != 0 {
+		klog.V(1).Infof("Using kubeconfig %q.", o.KubeconfigPath)
+		o.restConfig, err = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+			&clientcmd.ClientConfigLoadingRules{ExplicitPath: o.KubeconfigPath}, &clientcmd.ConfigOverrides{}).ClientConfig()
+		if err != nil {
+			return fmt.Errorf("can't create config from kubeConfigPath %q: %w", o.KubeconfigPath, err)
+		}
+	} else {
+		klog.V(1).Infof("No kubeconfig specified, using InClusterConfig.")
+		o.restConfig, err = restclient.InClusterConfig()
+		if err != nil {
+			return fmt.Errorf("can't create InClusterConfig: %w", err)
+		}
+	}
+
+	if len(o.TLSServerName) != 0 {
+		o.restConfig.TLSClientConfig.ServerName = o.TLSServerName
+	}
+
+	o.kubeClient, err = kubernetes.NewForConfig(o.restConfig)
+	if err != nil {
+		return fmt.Errorf("can't build kubernetes clientset: %w", err)
+	}
+
+	o.configClient, err = configeversionedclient.NewForConfig(o.restConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create config client: %w", err)
+	}
+
+	return nil
+}
+
+func (o *Options) Run() error {
+	var leWg sync.WaitGroup
+	var wg sync.WaitGroup
+	leCtx, leCancel := context.WithCancel(context.Background())
+	defer func() {
+		klog.Info("Waiting for controllers to finish...")
+		wg.Wait()
+
+		// Make sure we always flush, even on hard exit in the following path
+		klog.Flush()
+
+		leCancel()
+		klog.Info("Waiting for leaded election loop to finish...")
+		leWg.Wait()
+
+		klog.Flush()
+	}()
+
+	ctx, cancel := context.WithCancel(leCtx)
+	stopCh := server.SetupSignalHandler()
+	go func() {
+		<-stopCh
+		cancel()
+	}()
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return err
+	}
+	// add a uniquifier so that two processes on the same host don't accidentally both become active
+	id := hostname + "_" + string(uuid.NewUUID())
+	klog.V(4).Infof("Leaderelection ID is %q", id)
+
+	lock := &resourcelock.ConfigMapLock{
```

Comment:
pretty sure have leader election boilerplate in the operator command stuff. Also, I don't think that this should require leader election.  Race and have one winner.

Reply:
I'd prefer recovery rotation logs to be consistent, not scattered around between pods, so it is easier to reason about what happened. Migrated to library-go based LE.